test/src/loaders/workspace_uploader/workspace_uploader_test.py

```diff
@@ -1,5 +1,210 @@
+    )
+
+    expected_data = {
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    assert uploaded
+    assert expected_data == data
+
+
+def test_fetch_assemblies_to_upload(setup_and_teardown):
+    tmp_dir, assembly_dirs, assembly_names, _, workspace_id, upload_env_key, _ = setup_and_teardown
+    collection_source_dir = Path(tmp_dir) / "collectionssource"
+    upload_file_ext = ["genomic.fna.gz"]
+
+    count, wait_to_upload_assemblies = workspace_uploader._fetch_assemblies_to_upload(
+        upload_env_key, workspace_id, collection_source_dir, upload_file_ext
+    )
+    
+    expected_count = len(assembly_names)
+    expected_wait_to_upload_assemblies = {
+        assembly_name: assembly_dir
+        for assembly_name, assembly_dir in zip(assembly_names, assembly_dirs)
+    }
+
+    assert expected_count == count
+    assert expected_wait_to_upload_assemblies == wait_to_upload_assemblies
+
+
+def test_prepare_skd_job_dir_to_upload(setup_and_teardown):
+    tmp_dir, assembly_dirs, assembly_names, _, _, _, _ = setup_and_teardown
+    wait_to_upload_assemblies = {
+        assembly_name: assembly_dir
+        for assembly_name, assembly_dir in zip(assembly_names, assembly_dirs)
+    }
+
+    conf = Mock()
+    username = "kbase"
+    job_dir = loader_helper.make_job_dir(tmp_dir, username)
+    conf.job_data_dir = loader_helper.make_job_data_dir(job_dir)
+    data_dir = workspace_uploader._prepare_skd_job_dir_to_upload(conf, wait_to_upload_assemblies)
+    
+    assert sorted(os.listdir(data_dir)) == sorted(assembly_names)
+
+
+def test_post_process(setup_and_teardown):
+    tmp_dir, assembly_dirs, assembly_names, upas, workspace_id, upload_env_key, _ = setup_and_teardown
+    upload_dir = Path(tmp_dir) / "upload_dir"
+    upload_dir.mkdir()
+    output_dir = Path(tmp_dir) / "output_dir"
+    output_dir.mkdir()
+
+    upa = upas[0]
+    host_assembly_dir = assembly_dirs[0]
+    assembly_name = assembly_names[0]
+
+    workspace_uploader._post_process(
+        upload_env_key,
+        workspace_id,
+        host_assembly_dir,
+        assembly_name,
+        upload_dir,
+        output_dir,
+        upa,
+    )
+
+    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
+        upload_env_key, workspace_id, host_assembly_dir, assembly_name
+    )
+    expected_data = {
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    assert uploaded
+    assert expected_data == data
+    assert os.readlink(os.path.join(upload_dir, upa)) == os.path.join(output_dir, upa)
+
+
+def test_upload_assembly_to_workspace(setup_and_teardown):
+    _, _, assembly_names, upas, workspace_id, _, _  = setup_and_teardown
+    upa = upas[0]
+    assembly_name = assembly_names[0]
+    file_path = "/path/to/file/in/AssembilyUtil"
+
+    conf = Mock()
+    conf.asu.save_assembly_from_fasta2.return_value = {"upa": upa}
```

Comment:
I would use `create_autospec` here for `asu` rather than a bare mock
https://m1lt0n.github.io/python/unittest/mock/autospec/mocking-with-autospec/
https://docs.python.org/3/library/unittest.mock.html#auto-speccing

Reply:
Updated!

Reply:
You should  add `spec_set=True, instance=True` like this:
https://github.com/kbase/sample_service/blob/9ffd516a66d5dc046aa48a339560b158928eb2c0/test/core/workspace_test.py#L288

Reply:
Is there a workaround can help me get rid of AttributeError: Mock object has no attribute 'asu' error?

Reply:
Use a regular `Mock` for `Conf`. `create_autospec` doesn't know about dynamically created properties like `asu`

Reply:
So something like
```
conf = Mock()
conf.asu = create_autospec(...)
```

Reply:
works like a charm. Updated!

Reply:
:+1: